Dart/src/com/jetbrains/lang/dart/util/DotPackagesFileUtil.java

```diff
@@ -15,15 +19,149 @@
+    Pair<Long, Map<String, String>> data = packageConfigJsonFile.getUserData(MOD_STAMP_TO_PACKAGES_MAP);
+
+    final Long currentTimestamp = packageConfigJsonFile.getModificationCount();
+    final Long cachedTimestamp = Pair.getFirst(data);
+
+    if (cachedTimestamp == null || !cachedTimestamp.equals(currentTimestamp)) {
+      data = null;
+      packageConfigJsonFile.putUserData(MOD_STAMP_TO_PACKAGES_MAP, null);
+      final Map<String, String> packagesMap = packagesMapFromJson(packageConfigJsonFile);
+
+      if (packagesMap != null) {
+        data = Pair.create(currentTimestamp, packagesMap);
+        packageConfigJsonFile.putUserData(MOD_STAMP_TO_PACKAGES_MAP, data);
+      }
+    }
+
+    return Pair.getSecond(data);
+  }
+
+  /**
+   * Example JSON parsed:
+   * ```
+   * {
+   * "configVersion": 2,
+   * "packages": [
+   * {
+   * "name": "pedantic",
+   * "rootUri": "file:///Users/jwren/.pub-cache/hosted/pub.dartlang.org/pedantic-1.11.0",
+   * "packageUri": "lib/",
+   * "languageVersion": "2.12"
+   * },
+   * {
+   * "name": "console_dart",
+   * "rootUri": "../",
+   * "packageUri": "lib/",
+   * "languageVersion": "2.12"
+   * }
+   * ],
+   * "generated": "2022-03-04T22:10:17.132325Z",
+   * "generator": "pub",
+   * "generatorVersion": "2.17.0-edge.f9147d933ef019c7e304c19ac039f57226ce1e37"
+   * }
+   * ```
+   */
+  @Nullable
+  private static Map<String, String> packagesMapFromJson(@NotNull final VirtualFile packageConfigJsonFile) {
+    String fileContentsStr = FileUtil.loadFileOrNull(packageConfigJsonFile.getPath());
+    if (fileContentsStr == null) {
+      return null;
+    }
+
+    final JsonElement jsonElement = JsonParser.parseString(fileContentsStr);
+    if (jsonElement instanceof JsonObject &&
+        ((JsonObject)jsonElement).get("packages") != null &&
+        ((JsonObject)jsonElement).get("packages").isJsonArray()) {
+      final Map<String, String> result = new HashMap<>();
+      final JsonArray jsonArray = ((JsonObject)jsonElement).get("packages").getAsJsonArray();
+      for (JsonElement element : jsonArray) {
+        JsonObject jsonObjectPackage = element.getAsJsonObject();
+        if (jsonObjectPackage.get("name") != null &&
+            jsonObjectPackage.get("rootUri") != null &&
+            jsonObjectPackage.get("packageUri") != null
+        ) {
+          final String packageName = jsonObjectPackage.get("name").getAsString();
+          final String rootUriValue = jsonObjectPackage.get("rootUri").getAsString();
+          final String packageUriValue = jsonObjectPackage.get("packageUri").getAsString();
+          // need to protect '+' chars because URLDecoder.decode replaces '+' with space
+          final String encodedUriWithoutPluses = StringUtil.replace(rootUriValue + "/" + packageUriValue, "+", "%2B");
+          final String uri = URLUtil.decode(encodedUriWithoutPluses);
+          final String packageUri = getAbsolutePackageRootPath(packageConfigJsonFile.getParent().getParent(), uri);
```

Comment:
@jwren Not sure `packageConfigJsonFile.getParent().getParent()` is correct here. See https://youtrack.jetbrains.com/issue/WEB-55363. I guess it should be `packageConfigJsonFile.getParent()`.